File: advanced_benchmark.py
#!/usr/bin/env python3
"""
Advanced Benchmark Suite for Anoverif
Analyzes timing variations and request performance patterns
"""

import asyncio
import aiohttp
import time
import json
import statistics
import argparse
import sys
import matplotlib.pyplot as plt
import numpy as np
from dataclasses import dataclass
from typing import List, Dict, Any
from concurrent.futures import ThreadPoolExecutor
import threading
import random


@dataclass
class RequestResult:
    """Individual request result with detailed timing"""

    request_id: int
    start_time: float
    end_time: float
    response_time: float
    status_code: int
    success: bool
    result_value: bool
    error: str = None
    queue_time: float = 0.0  # Time in queue before processing
    processing_time: float = 0.0  # Actual processing time


class BenchmarkSuite:
    def __init__(self, base_url: str = "http://localhost:8082/verify"):
        self.base_url = base_url
        self.results: List[RequestResult] = []
        self.session = None

    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()

    async def send_request(self, request_id: int, idval: str) -> RequestResult:
        """Send individual request with detailed timing"""
        start_time = time.time()

        try:
            payload = {"idval": idval}
            async with self.session.post(
                self.base_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=aiohttp.ClientTimeout(total=10),
            ) as response:
                end_time = time.time()
                response_time = (end_time - start_time) * 1000  # Convert to ms

                response_data = await response.json()

                return RequestResult(
                    request_id=request_id,
                    start_time=start_time,
                    end_time=end_time,
                    response_time=response_time,
                    status_code=response.status,
                    success=response_data.get("success", False),
                    result_value=response_data.get("result", False),
                    error=response_data.get("error"),
                )

        except Exception as e:
            end_time = time.time()
            response_time = (end_time - start_time) * 1000

            return RequestResult(
                request_id=request_id,
                start_time=start_time,
                end_time=end_time,
                response_time=response_time,
                status_code=0,
                success=False,
                result_value=False,
                error=str(e),
            )

    async def run_burst_test(
        self, burst_size: int, burst_interval: float
    ) -> List[RequestResult]:
        """Run burst test with specified burst size and interval"""
        print(f"🚀 Running burst test: {burst_size} requests every {burst_interval}s")

        results = []
        burst_count = 0

        for burst in range(5):  # 5 bursts
            burst_start = time.time()
            print(f"  Burst {burst + 1}/5 starting...")

            # Create tasks for this burst
            tasks = []
            for i in range(burst_size):
                request_id = burst * burst_size + i
                idval = f"burst_{burst}_req_{i}_{random.randint(1000, 9999)}"
                tasks.append(self.send_request(request_id, idval))

            # Execute burst
            burst_results = await asyncio.gather(*tasks)
            results.extend(burst_results)

            burst_end = time.time()
            burst_duration = burst_end - burst_start
            print(f"  Burst {burst + 1} completed in {burst_duration:.2f}s")

            # Wait for next burst
            if burst < 4:  # Don't wait after last burst
                await asyncio.sleep(burst_interval)

        return results

    async def run_ramp_test(
        self, start_rps: int, end_rps: int, duration: int
    ) -> List[RequestResult]:
        """Run ramp test with increasing request rate"""
        print(f"📈 Running ramp test: {start_rps} to {end_rps} RPS over {duration}s")

        results = []
        start_time = time.time()
        request_id = 0

        while time.time() - start_time < duration:
            elapsed = time.time() - start_time
            progress = elapsed / duration
            current_rps = start_rps + (end_rps - start_rps) * progress

            # Calculate delay for current RPS
            delay = 1.0 / current_rps if current_rps > 0 else 1.0

            # Send request
            idval = f"ramp_req_{request_id}_{random.randint(1000, 9999)}"
            result = await self.send_request(request_id, idval)
            results.append(result)

            request_id += 1

            # Progress indicator
            if request_id % 10 == 0:
                print(
                    f"  Progress: {progress*100:.1f}% - Current RPS: {current_rps:.1f}"
                )

            await asyncio.sleep(delay)

        return results

    async def run_sustained_test(self, rps: int, duration: int) -> List[RequestResult]:
        """Run sustained load test"""
        print(f"⚡ Running sustained test: {rps} RPS for {duration}s")

        results = []
        start_time = time.time()
        request_id = 0
        delay = 1.0 / rps

        while time.time() - start_time < duration:
            idval = f"sustained_req_{request_id}_{random.randint(1000, 9999)}"
            result = await self.send_request(request_id, idval)
            results.append(result)

            request_id += 1

            if request_id % 20 == 0:
                elapsed = time.time() - start_time
                print(f"  Progress: {elapsed:.1f}s - Requests sent: {request_id}")

            await asyncio.sleep(delay)

        return results

    async def run_concurrent_test(
        self, concurrency: int, total_requests: int
    ) -> List[RequestResult]:
        """Run concurrent test with specified concurrency level"""
        print(
            f"🔀 Running concurrent test: {concurrency} concurrent, {total_requests} total"
        )

        semaphore = asyncio.Semaphore(concurrency)

        async def limited_request(request_id: int, idval: str):
            async with semaphore:
                return await self.send_request(request_id, idval)

        tasks = []
        for i in range(total_requests):
            idval = f"concurrent_req_{i}_{random.randint(1000, 9999)}"
            tasks.append(limited_request(i, idval))

        return await asyncio.gather(*tasks)


def analyze_results(results: List[RequestResult]) -> Dict[str, Any]:
    """Analyze benchmark results and generate statistics"""
    if not results:
        return {}

    # Filter successful requests
    successful = [r for r in results if r.success]
    failed = [r for r in results if not r.success]

    # Response times
    response_times = [r.response_time for r in successful]

    if not response_times:
        return {"error": "No successful requests"}

    # Basic statistics
    stats = {
        "total_requests": len(results),
        "successful_requests": len(successful),
        "failed_requests": len(failed),
        "success_rate": len(successful) / len(results) * 100,
        "response_times": {
            "min": min(response_times),
            "max": max(response_times),
            "mean": statistics.mean(response_times),
            "median": statistics.median(response_times),
            "p95": np.percentile(response_times, 95),
            "p99": np.percentile(response_times, 99),
            "std_dev": (
                statistics.stdev(response_times) if len(response_times) > 1 else 0
            ),
        },
        "timing_analysis": analyze_timing_patterns(successful),
        "error_analysis": analyze_errors(failed),
    }

    return stats


def analyze_timing_patterns(results: List[RequestResult]) -> Dict[str, Any]:
    """Analyze timing patterns and variations"""
    if len(results) < 2:
        return {}

    times = [r.response_time for r in results]
    timestamps = [r.start_time for r in results]

    # Time series analysis
    time_series = list(zip(timestamps, times))
    time_series.sort(key=lambda x: x[0])

    # Calculate variations
    variations = []
    for i in range(1, len(time_series)):
        variation = abs(time_series[i][1] - time_series[i - 1][1])
        variations.append(variation)

    # Detect patterns
    patterns = {
        "max_variation": max(variations) if variations else 0,
        "avg_variation": statistics.mean(variations) if variations else 0,
        "timing_consistency": (
            1 - (statistics.stdev(times) / statistics.mean(times))
            if statistics.mean(times) > 0
            else 0
        ),
        "outliers": detect_outliers(times),
        "trending": analyze_trend(times),
    }

    return patterns


def detect_outliers(data: List[float], threshold: float = 2.0) -> Dict[str, Any]:
    """Detect outliers using z-score method"""
    if len(data) < 3:
        return {"count": 0, "percentage": 0}

    mean = statistics.mean(data)
    std_dev = statistics.stdev(data)

    outliers = []
    for i, value in enumerate(data):
        z_score = abs(value - mean) / std_dev if std_dev > 0 else 0
        if z_score > threshold:
            outliers.append({"index": i, "value": value, "z_score": z_score})

    return {
        "count": len(outliers),
        "percentage": len(outliers) / len(data) * 100,
        "outliers": outliers[:10],  # Top 10 outliers
    }


def analyze_trend(data: List[float]) -> str:
    """Analyze if there's a trending pattern in response times"""
    if len(data) < 10:
        return "insufficient_data"

    # Simple linear regression to detect trend
    x = list(range(len(data)))
    y = data

    n = len(data)
    sum_x = sum(x)
    sum_y = sum(y)
    sum_xy = sum(xi * yi for xi, yi in zip(x, y))
    sum_x2 = sum(xi * xi for xi in x)

    slope = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)

    if slope > 0.1:
        return "increasing"
    elif slope < -0.1:
        return "decreasing"
    else:
        return "stable"


def analyze_errors(failed_results: List[RequestResult]) -> Dict[str, Any]:
    """Analyze error patterns"""
    if not failed_results:
        return {"total": 0, "types": {}}

    error_types = {}
    for result in failed_results:
        error = result.error or "unknown"
        error_types[error] = error_types.get(error, 0) + 1

    return {
        "total": len(failed_results),
        "types": error_types,
        "most_common": (
            max(error_types.items(), key=lambda x: x[1]) if error_types else None
        ),
    }


def generate_report(test_name: str, stats: Dict[str, Any]) -> str:
    """Generate detailed report"""
    if "error" in stats:
        return f"❌ {test_name}: {stats['error']}"

    rt = stats["response_times"]
    ta = stats["timing_analysis"]

    report = f"""
🎯 {test_name} Results:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

📊 Basic Metrics:
  • Total Requests: {stats['total_requests']}
  • Success Rate: {stats['success_rate']:.1f}%
  • Failed Requests: {stats['failed_requests']}

⏱️ Response Time Analysis:
  • Min: {rt['min']:.2f}ms
  • Max: {rt['max']:.2f}ms  
  • Mean: {rt['mean']:.2f}ms
  • Median: {rt['median']:.2f}ms
  • 95th Percentile: {rt['p95']:.2f}ms
  • 99th Percentile: {rt['p99']:.2f}ms
  • Std Deviation: {rt['std_dev']:.2f}ms

📈 Timing Patterns:
  • Max Variation: {ta.get('max_variation', 0):.2f}ms
  • Avg Variation: {ta.get('avg_variation', 0):.2f}ms
  • Consistency Score: {ta.get('timing_consistency', 0):.3f}
  • Trend: {ta.get('trending', 'unknown')}
  • Outliers: {ta.get('outliers', {}).get('count', 0)} ({ta.get('outliers', {}).get('percentage', 0):.1f}%)
"""

    if stats.get("error_analysis", {}).get("total", 0) > 0:
        ea = stats["error_analysis"]
        report += f"""
❌ Error Analysis:
  • Total Errors: {ea['total']}
  • Most Common: {ea.get('most_common', ['unknown', 0])[0]} ({ea.get('most_common', ['unknown', 0])[1]} times)
"""

    return report


async def main():
    parser = argparse.ArgumentParser(description="Advanced Anoverif Benchmark Suite")
    parser.add_argument(
        "--url", default="http://localhost:8082/verify", help="Server URL"
    )
    parser.add_argument("--all", action="store_true", help="Run all benchmark tests")
    parser.add_argument("--burst", action="store_true", help="Run burst test")
    parser.add_argument("--ramp", action="store_true", help="Run ramp test")
    parser.add_argument("--sustained", action="store_true", help="Run sustained test")
    parser.add_argument("--concurrent", action="store_true", help="Run concurrent test")
    parser.add_argument("--output", help="Output file for detailed results")

    args = parser.parse_args()

    if not any([args.all, args.burst, args.ramp, args.sustained, args.concurrent]):
        args.all = True

    print("🚀 Advanced Anoverif Benchmark Suite")
    print("=" * 50)

    async with BenchmarkSuite(args.url) as benchmark:
        all_results = {}

        if args.all or args.burst:
            print("\n" + "=" * 50)
            results = await benchmark.run_burst_test(burst_size=20, burst_interval=2.0)
            stats = analyze_results(results)
            all_results["burst_test"] = {"results": results, "stats": stats}
            print(generate_report("Burst Test", stats))

        if args.all or args.ramp:
            print("\n" + "=" * 50)
            results = await benchmark.run_ramp_test(
                start_rps=5, end_rps=25, duration=30
            )
            stats = analyze_results(results)
            all_results["ramp_test"] = {"results": results, "stats": stats}
            print(generate_report("Ramp Test", stats))

        if args.all or args.sustained:
            print("\n" + "=" * 50)
            results = await benchmark.run_sustained_test(rps=15, duration=20)
            stats = analyze_results(results)
            all_results["sustained_test"] = {"results": results, "stats": stats}
            print(generate_report("Sustained Test", stats))

        if args.all or args.concurrent:
            print("\n" + "=" * 50)
            results = await benchmark.run_concurrent_test(
                concurrency=30, total_requests=150
            )
            stats = analyze_results(results)
            all_results["concurrent_test"] = {"results": results, "stats": stats}
            print(generate_report("Concurrent Test", stats))

        # Save detailed results if requested
        if args.output:
            with open(args.output, "w") as f:
                json.dump(all_results, f, indent=2, default=str)
            print(f"\n💾 Detailed results saved to {args.output}")

        print("\n🎉 Benchmark suite completed!")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n⚠️ Benchmark interrupted by user")
    except Exception as e:
        print(f"\n❌ Benchmark failed: {e}")
        sys.exit(1)

File: src/advanced_benchmark.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <thread>
#include <future>
#include <random>
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <curl/curl.h>
#include <json/json.h>
#include <atomic>
#include <mutex>

namespace anoverif
{
    struct BenchmarkResult
    {
        int request_id;
        double start_time_ms;
        double end_time_ms;
        double response_time_ms;
        bool success;
        bool result_value;
        std::string error;
    };

    struct BenchmarkStats
    {
        int total_requests = 0;
        int successful_requests = 0;
        int failed_requests = 0;
        double success_rate = 0.0;
        double min_response_time = 0.0;
        double max_response_time = 0.0;
        double avg_response_time = 0.0;
        double median_response_time = 0.0;
        double p95_response_time = 0.0;
        double p99_response_time = 0.0;
        double std_deviation = 0.0;
        double max_variation = 0.0;
        double avg_variation = 0.0;
        std::vector<double> response_times;
    };

    class AdvancedBenchmark
    {
    private:
        std::string server_url_;
        std::vector<BenchmarkResult> results_;
        std::mutex results_mutex_;
        std::atomic<int> request_counter_{0};

        static size_t WriteCallback(void *contents, size_t size, size_t nmemb, std::string *userp)
        {
            size_t total_size = size * nmemb;
            userp->append(static_cast<char *>(contents), total_size);
            return total_size;
        }

        BenchmarkResult send_request(int request_id, const std::string &idval)
        {
            BenchmarkResult result;
            result.request_id = request_id;
            result.start_time_ms = get_current_time_ms();

            CURL *curl = curl_easy_init();
            std::string response_data;

            if (curl)
            {
                // Create JSON payload
                Json::Value payload;
                payload["idval"] = idval;
                Json::StreamWriterBuilder builder;
                std::string json_data = Json::writeString(builder, payload);

                // Setup headers
                struct curl_slist *headers = nullptr;
                headers = curl_slist_append(headers, "Content-Type: application/json");

                // Configure CURL
                curl_easy_setopt(curl, CURLOPT_URL, server_url_.c_str());
                curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_data.c_str());
                curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
                curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
                curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);
                curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);

                // Execute request
                CURLcode res = curl_easy_perform(curl);
                result.end_time_ms = get_current_time_ms();
                result.response_time_ms = result.end_time_ms - result.start_time_ms;

                if (res == CURLE_OK)
                {
                    // Parse response
                    Json::Value response_json;
                    Json::Reader reader;
                    if (reader.parse(response_data, response_json))
                    {
                        result.success = response_json.get("success", false).asBool();
                        result.result_value = response_json.get("result", false).asBool();
                    }
                    else
                    {
                        result.success = false;
                        result.error = "Failed to parse JSON response";
                    }
                }
                else
                {
                    result.success = false;
                    result.error = curl_easy_strerror(res);
                }

                curl_slist_free_all(headers);
                curl_easy_cleanup(curl);
            }
            else
            {
                result.end_time_ms = get_current_time_ms();
                result.response_time_ms = result.end_time_ms - result.start_time_ms;
                result.success = false;
                result.error = "Failed to initialize CURL";
            }

            return result;
        }

        double get_current_time_ms()
        {
            auto now = std::chrono::high_resolution_clock::now();
            auto duration = now.time_since_epoch();
            return std::chrono::duration<double, std::milli>(duration).count();
        }

    public:
        AdvancedBenchmark(const std::string &server_url) : server_url_(server_url)
        {
            curl_global_init(CURL_GLOBAL_DEFAULT);
        }

        ~AdvancedBenchmark()
        {
            curl_global_cleanup();
        }

        // Burst test: Send batches of requests with intervals
        void run_burst_test(int burst_size, int num_bursts, double interval_seconds)
        {
            std::cout << "🚀 Running Burst Test: " << burst_size << " requests x " << num_bursts
                      << " bursts, " << interval_seconds << "s intervals" << std::endl;

            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<> dis(1000, 9999);

            for (int burst = 0; burst < num_bursts; ++burst)
            {
                std::cout << "  Burst " << (burst + 1) << "/" << num_bursts << "..." << std::endl;

                std::vector<std::future<BenchmarkResult>> futures;

                // Launch burst
                for (int i = 0; i < burst_size; ++i)
                {
                    int req_id = request_counter_++;
                    std::string idval = "burst_" + std::to_string(burst) + "_req_" +
                                        std::to_string(i) + "_" + std::to_string(dis(gen));

                    futures.push_back(std::async(std::launch::async,
                                                 [this, req_id, idval]()
                                                 {
                                                     return this->send_request(req_id, idval);
                                                 }));
                }

                // Collect results
                for (auto &future : futures)
                {
                    BenchmarkResult result = future.get();
                    std::lock_guard<std::mutex> lock(results_mutex_);
                    results_.push_back(result);
                }

                // Wait before next burst (except for last one)
                if (burst < num_bursts - 1)
                {
                    std::this_thread::sleep_for(std::chrono::duration<double>(interval_seconds));
                }
            }
        }

        // Ramp test: Gradually increase request rate
        void run_ramp_test(double start_rps, double end_rps, double duration_seconds)
        {
            std::cout << "📈 Running Ramp Test: " << start_rps << " to " << end_rps
                      << " RPS over " << duration_seconds << " seconds" << std::endl;

            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<> dis(1000, 9999);

            double start_time = get_current_time_ms();
            double end_time = start_time + (duration_seconds * 1000);

            while (get_current_time_ms() < end_time)
            {
                double elapsed = (get_current_time_ms() - start_time) / 1000.0;
                double progress = elapsed / duration_seconds;
                double current_rps = start_rps + (end_rps - start_rps) * progress;

                if (current_rps > 0)
                {
                    double delay_ms = 1000.0 / current_rps;

                    int req_id = request_counter_++;
                    std::string idval = "ramp_req_" + std::to_string(req_id) + "_" + std::to_string(dis(gen));

                    // Send request asynchronously
                    std::thread([this, req_id, idval]()
                                {
                        BenchmarkResult result = this->send_request(req_id, idval);
                        std::lock_guard<std::mutex> lock(this->results_mutex_);
                        this->results_.push_back(result); })
                        .detach();

                    // Progress indicator
                    if (req_id % 10 == 0)
                    {
                        std::cout << "  Progress: " << std::fixed << std::setprecision(1)
                                  << (progress * 100) << "% - Current RPS: "
                                  << std::setprecision(1) << current_rps << std::endl;
                    }

                    std::this_thread::sleep_for(std::chrono::duration<double, std::milli>(delay_ms));
                }
            }

            // Wait for remaining requests to complete
            std::this_thread::sleep_for(std::chrono::seconds(2));
        }

        // Sustained test: Constant request rate
        void run_sustained_test(double rps, double duration_seconds)
        {
            std::cout << "⚡ Running Sustained Test: " << rps << " RPS for "
                      << duration_seconds << " seconds" << std::endl;

            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<> dis(1000, 9999);

            double delay_ms = 1000.0 / rps;
            double start_time = get_current_time_ms();
            double end_time = start_time + (duration_seconds * 1000);

            while (get_current_time_ms() < end_time)
            {
                int req_id = request_counter_++;
                std::string idval = "sustained_req_" + std::to_string(req_id) + "_" + std::to_string(dis(gen));

                // Send request asynchronously
                std::thread([this, req_id, idval]()
                            {
                    BenchmarkResult result = this->send_request(req_id, idval);
                    std::lock_guard<std::mutex> lock(this->results_mutex_);
                    this->results_.push_back(result); })
                    .detach();

                if (req_id % 20 == 0)
                {
                    double elapsed = (get_current_time_ms() - start_time) / 1000.0;
                    std::cout << "  Progress: " << std::fixed << std::setprecision(1)
                              << elapsed << "s - Requests sent: " << req_id << std::endl;
                }

                std::this_thread::sleep_for(std::chrono::duration<double, std::milli>(delay_ms));
            }

            // Wait for remaining requests to complete
            std::this_thread::sleep_for(std::chrono::seconds(2));
        }

        // High concurrency test
        void run_high_concurrency_test(int concurrency, int total_requests)
        {
            std::cout << "🔀 Running High Concurrency Test: " << concurrency
                      << " concurrent, " << total_requests << " total requests" << std::endl;

            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<> dis(1000, 9999);

            std::vector<std::future<BenchmarkResult>> futures;
            std::atomic<int> active_requests{0};

            for (int i = 0; i < total_requests; ++i)
            {
                // Wait if we've reached concurrency limit
                while (active_requests >= concurrency)
                {
                    std::this_thread::sleep_for(std::chrono::milliseconds(1));
                }

                int req_id = request_counter_++;
                std::string idval = "concurrent_req_" + std::to_string(i) + "_" + std::to_string(dis(gen));

                active_requests++;
                futures.push_back(std::async(std::launch::async,
                                             [this, req_id, idval, &active_requests]()
                                             {
                                                 BenchmarkResult result = this->send_request(req_id, idval);
                                                 active_requests--;
                                                 return result;
                                             }));

                if (i % 50 == 0)
                {
                    std::cout << "  Progress: " << i << "/" << total_requests
                              << " requests queued" << std::endl;
                }
            }

            // Collect all results
            for (auto &future : futures)
            {
                BenchmarkResult result = future.get();
                std::lock_guard<std::mutex> lock(results_mutex_);
                results_.push_back(result);
            }
        }

        BenchmarkStats analyze_results()
        {
            BenchmarkStats stats;
            std::lock_guard<std::mutex> lock(results_mutex_);

            if (results_.empty())
            {
                return stats;
            }

            stats.total_requests = results_.size();

            // Separate successful and failed requests
            std::vector<BenchmarkResult> successful;
            std::vector<BenchmarkResult> failed;

            for (const auto &result : results_)
            {
                if (result.success)
                {
                    successful.push_back(result);
                    stats.response_times.push_back(result.response_time_ms);
                }
                else
                {
                    failed.push_back(result);
                }
            }

            stats.successful_requests = successful.size();
            stats.failed_requests = failed.size();
            stats.success_rate = (double)successful.size() / results_.size() * 100.0;

            if (!stats.response_times.empty())
            {
                // Sort for percentile calculations
                std::sort(stats.response_times.begin(), stats.response_times.end());

                stats.min_response_time = stats.response_times.front();
                stats.max_response_time = stats.response_times.back();

                // Calculate average
                double sum = 0;
                for (double time : stats.response_times)
                {
                    sum += time;
                }
                stats.avg_response_time = sum / stats.response_times.size();

                // Calculate median
                size_t mid = stats.response_times.size() / 2;
                if (stats.response_times.size() % 2 == 0)
                {
                    stats.median_response_time = (stats.response_times[mid - 1] + stats.response_times[mid]) / 2.0;
                }
                else
                {
                    stats.median_response_time = stats.response_times[mid];
                }

                // Calculate percentiles
                stats.p95_response_time = stats.response_times[static_cast<size_t>(stats.response_times.size() * 0.95)];
                stats.p99_response_time = stats.response_times[static_cast<size_t>(stats.response_times.size() * 0.99)];

                // Calculate standard deviation
                double variance = 0;
                for (double time : stats.response_times)
                {
                    variance += (time - stats.avg_response_time) * (time - stats.avg_response_time);
                }
                stats.std_deviation = std::sqrt(variance / stats.response_times.size());

                // Calculate variations
                std::vector<double> variations;
                for (size_t i = 1; i < stats.response_times.size(); ++i)
                {
                    variations.push_back(std::abs(stats.response_times[i] - stats.response_times[i - 1]));
                }

                if (!variations.empty())
                {
                    stats.max_variation = *std::max_element(variations.begin(), variations.end());
                    double sum_var = 0;
                    for (double var : variations)
                    {
                        sum_var += var;
                    }
                    stats.avg_variation = sum_var / variations.size();
                }
            }

            return stats;
        }

        void print_detailed_report(const std::string &test_name, const BenchmarkStats &stats)
        {
            std::cout << "\n🎯 " << test_name << " Results:" << std::endl;
            std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;

            std::cout << "\n📊 Basic Metrics:" << std::endl;
            std::cout << "  • Total Requests: " << stats.total_requests << std::endl;
            std::cout << "  • Successful: " << stats.successful_requests << std::endl;
            std::cout << "  • Failed: " << stats.failed_requests << std::endl;
            std::cout << "  • Success Rate: " << std::fixed << std::setprecision(1)
                      << stats.success_rate << "%" << std::endl;

            if (stats.successful_requests > 0)
            {
                std::cout << "\n⏱️ Response Time Analysis:" << std::endl;
                std::cout << "  • Min: " << std::fixed << std::setprecision(2)
                          << stats.min_response_time << "ms" << std::endl;
                std::cout << "  • Max: " << std::setprecision(2)
                          << stats.max_response_time << "ms" << std::endl;
                std::cout << "  • Average: " << std::setprecision(2)
                          << stats.avg_response_time << "ms" << std::endl;
                std::cout << "  • Median: " << std::setprecision(2)
                          << stats.median_response_time << "ms" << std::endl;
                std::cout << "  • 95th Percentile: " << std::setprecision(2)
                          << stats.p95_response_time << "ms" << std::endl;
                std::cout << "  • 99th Percentile: " << std::setprecision(2)
                          << stats.p99_response_time << "ms" << std::endl;
                std::cout << "  • Std Deviation: " << std::setprecision(2)
                          << stats.std_deviation << "ms" << std::endl;

                std::cout << "\n📈 Timing Variations:" << std::endl;
                std::cout << "  • Max Variation: " << std::setprecision(2)
                          << stats.max_variation << "ms" << std::endl;
                std::cout << "  • Avg Variation: " << std::setprecision(2)
                          << stats.avg_variation << "ms" << std::endl;

                double consistency = 1.0 - (stats.std_deviation / stats.avg_response_time);
                std::cout << "  • Consistency Score: " << std::setprecision(3)
                          << consistency << std::endl;
            }
        }

        void save_detailed_results(const std::string &filename)
        {
            std::ofstream file(filename);
            if (!file.is_open())
            {
                std::cerr << "Error: Could not open " << filename << " for writing" << std::endl;
                return;
            }

            file << "request_id,start_time_ms,end_time_ms,response_time_ms,success,result_value,error\n";

            std::lock_guard<std::mutex> lock(results_mutex_);
            for (const auto &result : results_)
            {
                file << result.request_id << ","
                     << std::fixed << std::setprecision(3) << result.start_time_ms << ","
                     << result.end_time_ms << ","
                     << result.response_time_ms << ","
                     << (result.success ? "true" : "false") << ","
                     << (result.result_value ? "true" : "false") << ","
                     << "\"" << result.error << "\"\n";
            }

            std::cout << "💾 Detailed results saved to " << filename << std::endl;
        }

        void clear_results()
        {
            std::lock_guard<std::mutex> lock(results_mutex_);
            results_.clear();
            request_counter_ = 0;
        }
    };
}

int main(int argc, char *argv[])
{
    std::string server_url = "http://localhost:8082/verify";

    if (argc > 1)
    {
        server_url = argv[1];
    }

    std::cout << "🚀 Advanced Anoverif Benchmark Suite" << std::endl;
    std::cout << "Server URL: " << server_url << std::endl;
    std::cout << std::string(50, '=') << std::endl;

    anoverif::AdvancedBenchmark benchmark(server_url);

    try
    {
        // Test 1: Burst Test
        std::cout << "\n"
                  << std::string(50, '=') << std::endl;
        benchmark.run_burst_test(15, 4, 2.0); // 15 requests x 4 bursts, 2s intervals
        auto stats = benchmark.analyze_results();
        benchmark.print_detailed_report("Burst Test", stats);
        benchmark.save_detailed_results("burst_test_results.csv");
        benchmark.clear_results();

        // Test 2: Ramp Test
        std::cout << "\n"
                  << std::string(50, '=') << std::endl;
        benchmark.run_ramp_test(5.0, 20.0, 25.0); // 5 to 20 RPS over 25 seconds
        stats = benchmark.analyze_results();
        benchmark.print_detailed_report("Ramp Test", stats);
        benchmark.save_detailed_results("ramp_test_results.csv");
        benchmark.clear_results();

        // Test 3: Sustained Test
        std::cout << "\n"
                  << std::string(50, '=') << std::endl;
        benchmark.run_sustained_test(12.0, 20.0); // 12 RPS for 20 seconds
        stats = benchmark.analyze_results();
        benchmark.print_detailed_report("Sustained Test", stats);
        benchmark.save_detailed_results("sustained_test_results.csv");
        benchmark.clear_results();

        // Test 4: High Concurrency Test
        std::cout << "\n"
                  << std::string(50, '=') << std::endl;
        benchmark.run_high_concurrency_test(25, 200); // 25 concurrent, 200 total
        stats = benchmark.analyze_results();
        benchmark.print_detailed_report("High Concurrency Test", stats);
        benchmark.save_detailed_results("concurrency_test_results.csv");

        std::cout << "\n🎉 Advanced benchmark suite completed!" << std::endl;
    }
    catch (const std::exception &e)
    {
        std::cerr << "❌ Benchmark failed: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
